### PR TITLE
[MPM] Neighbour search

### PR DESCRIPTION
**Pull request by peterjwilson**
Hi @KratosMultiphysics/mpm , the latest creation of the 'Ridgy Didgers' (Tobi and I) is to add the option of a neighbour-based search in the search element utility. 

### Overview

It works as follows:
- For the first timestep only, determine which geometries are next to each other and store this on the geometries (expensive, but one time only).
- For all timesteps, check if the MP is in the same element (very fast). If not, check if the MP is in a neighbouring element (also fast). If not, do the traditional bin-based search for that MP.

For problems with many background grid elements and few MPs involving a small number of timesteps the neighbour search method will probably be slower than the original.

For all other situations the neighbour search is expected to perform faster.

### Example 1 - Explicit cantilever beam with 3D tet grid. 
GiD output every 25 steps
MPs | 1440
Background grid elements | 8640

image

### Example 2 - Axisymmetric disk 2D tri grid. 
GiD output every 10 steps
MPs | 20
Background grid elements | 48

image

### CODE REVIEW

**Comment by tteschemacher** [comment]
Thank you @PeterJosephWilson for the speed comparisons. That's really sick.

If you, @KratosMultiphysics/mpm agree we move forward to only use the new search scheme?

**Comment by peterjwilson** [comment]
Previously, the neighbours of all elements were determined at the start of the sim. This latest commit only constructs geometry neighbours as needed, reducing the initial CPU cost and memory (especially for large grids).
- timestep n-1: MP was in geometry
- timestep n: MP no longer in geometry, determine neighbour geometries and search them

**Comment by RiccardoRossi** [comment]
cool!

just as a side note, take into account that you could speed up the speed of bin search (at the price of memory) by reducing the cell size. Still your methos is probably a better fit for MPM, so i am commenting this only for completenesss

**Comment by RiccardoRossi** [comment]
cool!

just as a side note, take into account that you could speed up the speed of bin search (at the price of memory) by reducing the cell size. Still your methos is probably a better fit for MPM, so i am commenting this only for completenesss

**Comment by tteschemacher** [comment]
@KratosMultiphysics/mpm are you fine with merging this?

File: applications/ParticleMechanicsApplication/python_scripts/mpm_explicit_solver.py
Comment (philbucher) [inline applications/ParticleMechanicsApplication/python_scripts/mpm_explicit_solver.py:29]: This sounds like it can/should be done with a process

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h
Comment (philbucher) [inline applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h]: please do remove delicate commented code like this (if it is intended!)

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h
Comment (philbucher) [inline applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h:78]: cannot you use the process from the core for this?

Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h:78]: which one are you referring to?

Comment (philbucher) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h:78]: one of those:
image

maybe @RiccardoRossi has a better overview of which one to use

Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h:78]: the ones in the core are either referring to elements/ conditions or nodes. This one is for geometries. Maybe we need to consider extending the core processes to geometries.

File: applications/ParticleMechanicsApplication/python_scripts/mpm_explicit_solver.py
Comment (peterjwilson) [reply applications/ParticleMechanicsApplication/python_scripts/mpm_explicit_solver.py:29]: This is just adding the option to disable or enable the explicit MP on edge issue. Although it may be 'processy', I think it's better to keep it in the search utility because this is exactly where the problem is detected, and the fix for the problem requires additional searches (and can use the search structure already in scope) .

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h
Comment (peterjwilson) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h]: dunzo 🚀

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h
Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h:78]: Hey, we had a deeper look at the processes and it seems that they're not suitable at the moment. We would require deeper adaption which would involve certain add ons in the core which we would like to postpone to future PRs.

**Comment by peterjwilson** [comment]
ping @philbucher, @tteschemacher, @KratosMultiphysics/mpm are you guys fine with merging this?

**Review by tteschemacher — APPROVED** [review]
I'm totally fine with it. Thanks a lot.

Let's wait until tomorrow for the merge to give Europe a chance.

**Review by philbucher — APPROVED** [review]
Haha cheers for waiting 

All good from my side

**Comment by antonialarese** [comment]
Better late than never (and before tomorrow meeting or today meeting (depending on the continent!)).
Good idea and very good work. Super efficient as usual. Thanks!

**Pull request merged**
